File: libs/daemon/client/src/mill/client/lock/MemoryLock.java
package mill.client.lock;

import java.util.concurrent.locks.ReentrantLock;

public class MemoryLock extends Lock {

  private final ReentrantLock innerLock = new ReentrantLock();

  @Override
  public boolean probe() {
    return !innerLock.isLocked();
  }

  @Override
  public Locked lock() {
    innerLock.lock();
    return new MemoryLocked(innerLock);
  }

  @Override
  public MemoryTryLocked tryLock() {
    MemoryTryLocked res =
        innerLock.tryLock() ? new MemoryTryLocked(innerLock) : new MemoryTryLocked(null);

    return res;
  }

  @Override
  public void close() throws Exception {
    try {
      innerLock.unlock();
    } catch (Exception e) {
      // If the lock is not currently locked, close() does nothing, but shouldn't throw
    }
  }
}

File: libs/daemon/client/src/mill/client/lock/MemoryLock.scala
package mill.client.lock

import java.util.concurrent.locks.ReentrantLock

class MemoryLock extends Lock {
  private val innerLock = new ReentrantLock()

  override def probe(): Boolean = !innerLock.isLocked

  override def lock(): Locked = {
    innerLock.lock()
    new MemoryLocked(innerLock)
  }

  override def tryLock(): TryLocked = {
    if (innerLock.tryLock()) new MemoryTryLocked(innerLock)
    else new MemoryTryLocked(null)
  }

  override def close(): Unit = {
    try innerLock.unlock()
    catch {
      case _: Exception =>
      // If the lock is not currently locked, close() does nothing
    }
  }
}

private[lock] class MemoryLocked(lock: java.util.concurrent.locks.Lock) extends Locked {
  override def release(): Unit = lock.unlock()
}

private[lock] class MemoryTryLocked(lock: java.util.concurrent.locks.Lock)
    extends MemoryLocked(lock) with TryLocked {
  override def isLocked: Boolean = lock != null
  override def release(): Unit = if (isLocked) super.release()
}
